### ISSUE TITLE
Trait refactoring

### ISSUE THREAD

**Issue by ishitatsuyuki** [rxRust/rxRust#62]
I believe traits used in rxRust contain a couple of anti-patterns and we should change some of those definitions.

1. `Observable` should be a trait. The current `struct Observable` should be renamed. (Even official ReactiveX libraries has `Observable` as an abstract class; so having it as a struct is pretty strange.)
  **EDIT**: Also, consider using associated type for `Observable`, otherwise it would cause trouble when used as a bound.
  It seems that what should be (the core of) `Observable` is currently called `RawSubscribable`. See below for additional *provided* methods that should be included in the `Observable` trait.
1. `RawSubscribable` is generic on Subscriber. It shouldn't be like this; it should be generic on `Item` and `Err` instead.
1. Operators are exposed as extension traits, and that's even different from how extension traits are commonly used. See an example from [Differential Dataflow](https://docs.rs/differential-dataflow/0.11.0/differential_dataflow/operators/reduce/trait.Reduce.html) as how extension traits are commonly implemented. Bounds should exist on the implementation, not the trait itself. Otherwise, it results in indecipherable generated API docs.
  Though, in the case of rxRust I believe we don't need them, as there's only one type that we are interested in: `Observable`. So we should just implement all of the operators as provided methods on the `Observable` trait. (This kind of design is more common, examples are `Iterator` & itertools and `Future`)
1. Maybe also consider eliminating the `Error` type on observables? Although this will make rxRust different from other implementations, a few crates, namely futures, has moved to opt-in error handling with `Result`, instead of `Item, Err` generic arguments.

**Comment by M-Adoo** [comment]
Thanks  give these good advice.  Some problem  and detail we should discuss.
1.  if `Observable` should be a trait,  how should methods to creating `Observable` return, like `interval`, `range` and so on.
2. I think so before.  but I removed `Item` and `Err`  from  `RawSubscribable`'s generic and replace with `Subscriber`,  that because when implement those operators, `Subscriber ` let we can split it into `Observer` and `Subscription`, and these more clear and  easier to implement a new operator.
3. I agree with 
> Bounds should exist on the implementation
and actually we has it before, but remove it for easy to finish  refactoring，we should add it back.
4. I think 
 > this will make rxRust different from other implementations
maybe a problem,  not like futures, ReactiveX has  standard API and people familiar with it.

**Comment by ishitatsuyuki** [comment]
> 1. if `Observable` should be a trait, how should methods creating `Observable` return, like interval, range and so on.

Well, just like how you implement operators? If you look at RxJava, you will see that they are implemented as a concrete class: [`ObservableInterval`](https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableInterval.java), [`ObservableRange`](https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableRange.java).

Although what is different from Java is that these types (classes) are often public instead of internal in Rust.

The methods creating these observables should be a "free function" (function that belongs to a module but not a struct).

> 2. I think so before. but I removed Item and Err from RawSubscribable's generic and replace with Subscriber, that because when implement those operators, Subscriber let we can split it into Observer and Subscription, and these more clear and easier to implement a new operator.

I don't see what you mean. If you have issues with English, you might use Chinese as I could read it.

The current signature doesn't really make sense, due to two reasons.

1. The `Subscriber` generic argument is always an instantiation of the `Subscriber<S, U>` type. If that's the case the generic argument should be at least `S, U`.
2. Instead of generic arguments, associated types should be used. Think: you can only have one implementation for `RawSubscribable`. In the current state you can have multiple, which isn't how it works and this will cause issues with type inference.

Also note that the definition of an observable is that `it has a subscribe function that takes an observer". It does not take a `Subscriber`, as you can see in any of the official implementations.

3. I mentioned two possibilities here: extension traits, and methods on the `Observable` trait itself. Have you consider the latter? Because I think it's the best fit for the use case of rxRust.

4. I know it's opinion based, so here's a clarification: it's a tradeoff between "API familiar for Rust ecosystem" and "API familiar for ReactiveX ecosystem". Which one to choose, I will leave it up to you.

**Comment by M-Adoo** [comment]
> Well, just like how you implement operators? If you look at RxJava, you will see that they are implemented as a concrete class: ObservableInterval, ObservableRange.
Although what is different from Java is that these types (classes) are often public instead of internal in Rust.The methods creating these observables should be a "free function" (function that belongs to a module but not a struct).

Yes, concrete class can do it.  But to implement a creating observable function we needn't implement a new struct/class, just a function is enough, and `rxjs` is in this way.  There is no reason to change to `concrete class` way.

>1. The Subscriber generic argument is always an instantiation of the Subscriber<S, U> type. If that's the case the generic argument should be at least S, U.
>2. Instead of generic arguments, associated types should be used. Think: you can only have one implementation for RawSubscribable. In the current state you can have multiple, which isn't how it works and this will cause issues with type inference.

1. ok, Subscriber<S, U>  is more clear. 
2. I agree it, a pr is welcome. 

>Also note that the definition of an observable is that it has a subscribe function that takes an observer". It does not take a Subscriber`, as you can see in any of the official implementations.

When user use rx, just need a subscribe function，but when implement a operator, both `rxjs` and `RxJava` accept  a Subscriber.

`rxjs` skip
```ts
export function skip<T>(count: number): MonoTypeOperatorFunction<T> {
  return (source: Observable<T>) => source.lift(new SkipOperator(count));
}

class SkipOperator<T> implements Operator<T, T> {
  constructor(private total: number) {
  }

  call(subscriber: Subscriber<T>, source: any): TeardownLogic {
    return source.subscribe(new SkipSubscriber(subscriber, this.total));
  }
}

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends {Ignored}
 */
class SkipSubscriber<T> extends Subscriber<T> {
  count: number = 0;

  constructor(destination: Subscriber<T>, private total: number) {
    super(destination);
  }

  protected _next(x: T) {
    if (++this.count > this.total) {
      this.destination.next(x);
    }
  }
}
```
RxJava Skip
```java
public final class FlowableSkip<T> extends AbstractFlowableWithUpstream<T, T> {
    final long n;
    public FlowableSkip(Flowable<T> source, long n) {
        super(source);
        this.n = n;
    }

    @Override
    protected void subscribeActual(Subscriber<? super T> s) {
        source.subscribe(new SkipSubscriber<>(s, n));
    }

    static final class SkipSubscriber<T> implements FlowableSubscriber<T>, Subscription {
        final Subscriber<? super T> downstream;
        long remaining;

        Subscription upstream;

        SkipSubscriber(Subscriber<? super T> actual, long n) {
            this.downstream = actual;
            this.remaining = n;
        }
       ...
    }
}
```
> 3. I mentioned two possibilities here: extension traits, and methods on the Observable trait itself. Have you consider the latter? Because I think it's the best fit for the use case of rxRust.

Rx has two many operators, extension traits let user can import what they want.

> 4. I know it's opinion based, so here's a clarification: it's a tradeoff between "API familiar for Rust ecosystem" and "API familiar for ReactiveX ecosystem". Which one to choose, I will leave it up to you.
 
Respect Rx here.

**Comment by ishitatsuyuki** [comment]
> Yes, concrete class can do it. But to implement a creating observable function we needn't implement a new struct/class, just a function is enough, and rxjs is in this way. There is no reason to change to concrete class way.

Implementations on dynamic-typed languages or interface-oriented languages does not need to expose these class because you don't need to know the concrete type in order to store it.

Also, when designing things as a part of (core) crate API, you should always expose the concrete types as public access. While impl Trait in return position is often convenient, it has its own limitations like not being able to store such types in structs, etc. (These might get solved once [existential types](https://github.com/rust-lang/rust/issues/63063) are stabilized, but for now, you need to expose as I described.)

Finally the point is that `Observable` should be a trait, otherwise that would be the biggest contradiction to ReactiveX's standard API. The current one could be renamed to `ObservableFromFn`.

> When user use rx, just need a subscribe function，but when implement a operator, both rxjs and RxJava accept a Subscriber.

The code you linked is from RxJava 1.x. In RxJava 2.x and 3.x `Subscriber` is gone. https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableSkip.java

The notion of subscription is basically replaced by `Disposable` in RxJava 2. I think we should follow that: using `Drop` feels much better than managing subscriptions manually in a RAII language like Rust.

> Rx has two many operators, extension traits let user can import what they want.

Why would you consider that an advantage? "let user import what they want" is only a useful idiom in JS because of tree shaking. In Rust (and probably Java) it's just more hassle because you need to add imports more frequently.

**Comment by M-Adoo** [comment]
I agree your opinions, I will  happy to merge a pr.

But for extension traits, I'll use a wasm version in my private project. And it's not so bad to use, user can easily import all operators : `use rxrust::ops::*;`.

**Comment by helgoboss** [comment]
First things first, @M-Adoo, thank you so much for publishing this library! Previously, the lack of a proper Rx implementation kept me from considering Rust for my purposes. Fortunately, your implementation looks very promising and I'm hoping to contribute to it. But before adding operators, I would like to understand a bit better some of your design decisions and gain more confidence in the soundness of rxRust's core.

When experimenting with rxRust and trying to write a new operator, I had similar questions in mind like @ishitatsuyuki. Infact, the first thing I was wondering was "How can I return something obversable from a function without exposing too much implementation detail?". I was desperately looking for an `Observable` trait but couldn't find it. So here my take on @ishitatsuyuki's points:

1. **Rename `RawSubscribable` to `Observable`:** Yes, please. I think this would avoid much confusion.
2. **`Observable` (= former `RawSubscribable`) generics overhaul:** Yes, it would be nice to be generic on observer and subscription, without having an intermediate `Subscriber` parameter. But it would perhaps be even better to be generic on just `Item` (and maybe `Err`), as @ishitatsuyuki suggested initially. I think this is what most Rx users would expect (I'm used to it from RxCpp, RxJava and rxjs used from TypeScript). Do you think it could be made generic on just `Item` and `Err`? Or did you choose another way because you were encountering a showstopper?
3. **Operators as provided methods on `Observable` instead of extension traits:** For all standard operators I would also prefer that. I think this improves things like API ergonomics, documentation, error messages and IDE auto-complete/navigation. About your last comment: I have not used wasm, so I don't know ... Is it really an advantage for Rust-wasm to have operators as extension traits (vs. provided methods)?
4. **Remove error type:** Currently no opinion

That brings me to the question: @ishitatsuyuki, are you working on any of these improvements already? Can I help?

In addition, I have the following question (I think it's closely related, so I hope it's okay to discuss it in the same GitHub issue):

5. What is the reason for not having more bounds on the generic type parameters of structs? Is it for improving type inference? For example,  `struct TakeObserver` has parameters `O` and `S`, but there are no bounds on them. `ObserverNext` implementation adds bounds (e.g. that `S` should be `SubscriptionLike`). I suspect that having bounds on struct type parameters would improve error messages, documentation and IDE auto-completion/navigation because of "fail fast". For example, now I get an error message that `subscribe_all()` doesn't exist for a certain operator struct. But the source of the problem is that the operator struct shouldn't even have been created with those type bounds in the first place.

**Comment by ishitatsuyuki** [comment]
> I have not used wasm, so I don't know ... Is it really an advantage for Rust-wasm to have operators as extension traits (vs. provided methods)?

No, these kind of things work in Rust via monomorphization. Generic methods are compiled in only if you actually call it.

> What is the reason for not having more bounds on the generic type parameters of structs?

@M-Adoo said "and actually we had it before, but I removed it to make it easier to finish refactoring. we should add it back."

And to get to the last question, I tried to get on refactoring but I'm not actively working on it due to the vast number of files I need to modify. (I'm currently using ReactiveX in another language to do what I want to do.) If you're going to work on it, I'd suggest that you comment out all operators and bring them back as you port them to the new generic signature and test them one by one.

**Comment by M-Adoo** [comment]
Thank you guys so much for contributing to this library and giving these valuable ideas.  I'll start to refactor and push a branch the next day or two, and glad to finish it together. But before coding, we should make these two points clear:

1. Observable trait 
> But it would perhaps be even better to be generic on just Item (and maybe Err), as @ishitatsuyuki suggested initially.

I tend to a RxJava 3.x style @ishitatsuyuki mentioned above.  Just accept an `Observer` type, neither `Item`, nor `Subscriber`.  `Observer` trait has associated type  `Item` and `Err`, operator should have  a self `Observer` implementation.  `Observable` trait define like this:

```rust
trait Observable {
   type O: Observer;
   type Unsub: SubscriptionLike;
   fn actual_subscribe(self, subscriber: O) -> Self::Unsub;
}
```

2. Extension traits
> > I have not used wasm, so I don't know ... Is it really an advantage for Rust-wasm to have operators as extension traits (vs. provided methods)?

> No, these kind of things work in Rust via monomorphization. Generic methods are compiled in only if you actually call it.

Generic methods generating specialized versions only if you actually call it. But I'm not sure if  trait methods like below will generating specialized versions: 
```rust 
pub trait Observable {
  fn first(self) -> FirstOp<Self>;
  fn last(self) -> LastOp<Self>;
  fn skip(self, n: u32) -> SkipOp<Self>;
  ...
}
```

**Comment by ishitatsuyuki** [comment]
> ```
> trait Observable {
>   type O: Observer;
>   type Unsub: SubscriptionLike;
>   fn actual_subscribe(self, subscriber: O) -> Self::Unsub;
> }
> ```

This is wrong. You might want to consult documentation on generics in Rust. The following is what should be reasonably implemented:

```
trait Observable {
   type Item;
   type Err;
   type Unsub; // RxJava style return values utilizes destructors
   fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self:Err>>(self, subscriber: O) -> Self::Unsub;
}
```

> Generic methods generating specialized versions only if you actually call it. But I'm not sure if trait methods like below will generating specialized versions:

Okay, I might should have worded that differently. The point is that whether you import a trait or not **does not affect binary size**. Things work very different from JavaScript tree shaking, and please stop assuming those patterns will provide any benefit.

**Comment by helgoboss** [comment]
> @M-Adoo said "and actually we had it before, but I removed it to make it easier to finish refactoring. we should add it back."

Right, I missed that. Makes sense.

> I'll start to refactor and push a branch the next day or two, and glad to finish it together.

Great, looking forward 👍

> ```rust
> trait Observable {
>   type Item;
>   type Err;
>   type Unsub; // RxJava style return values utilizes destructors
>   fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self:Err>>(self, subscriber: O) -> Self::Unsub;
> }
> ```

That sounds good. RxSwift does it the same way: If you merge [ObservableType](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/ObservableType.swift) and [ObservableConvertibleType](https://github.com/ReactiveX/RxSwift/blob/master/RxSwift/ObservableConvertibleType.swift), it looks essentially the same (slight difference: `Item` is named `Element` plus it's not generic on `Disposable`/`Unsub`/`Subscription` and `Err`). I mention RxSwift because Swift seems a bit closer to Rust than Java. It also has associated types for example.

**Comment by M-Adoo** [comment]
```rust
trait Observable {
   type Item;
   type Err;
   type Unsub; // RxJava style return values utilizes destructors
   fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self:Err>>(self, subscriber: O) -> Self::Unsub;
}
```
Looks good to me.  For subscription manage, We can implement `Drop` for `LocalSubscription` and 
`SharedSubscription`.

> Okay, I might should have worded that differently. The point is that whether you import a trait or not does not affect binary size. Things work very different from JavaScript tree shaking, and please stop assuming those patterns will provide any benefit.
I got it, forget  native code  optimize size in a different way,  compiled generating specialized versions but not in finally executable file.  Operators as provided methods on `Observable` is not a problem for me now.

Let's list things we need to do: 

1. [ ] `Observable` should split into  concrete type, not only use a `Observable` struct to wrap all.
     - [ ] rename `Observable` to `ObservableFromFn`
     - [ ] `Observable` creating functions return  concrete class like `ObservableRange `.
2.  [ ]  rename `RawSubscribable` to `Observable`, and change the trait define to 
```rust
 trait Observable {
  type Item;
  type Err;
  type Unsub; // RxJava style return values utilizes destructors
  fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self:Err>>(self, subscriber: O) -> Self::Unsub;
}
```
3. [ ] Add bounds to operator struct generics.
4. [ ] Use `Drop` to improve Subscription manage.
5. [ ]  Operators as provided methods on Observable instead of extension traits.

Do @ishitatsuyuki  and @helgoboss have anything to add ? I'll create a mr tomorrow, 
and feel free to pick some tasks you want to do.

**Comment by ishitatsuyuki** [comment]
I think that's a very good summary. Thanks!

You probably forgot that `Observer` trait also needs to be changed. It would be like this:

```rust
pub trait Observer {
    type Item;
    type Err;
    
    fn next(&mut self, value: Self::Item);
    fn error(&mut self, err: Self::Err);
    fn complete(&mut self);
}
```

**Comment by M-Adoo** [comment]
> I think that's a very good summary. Thanks!
> 
> You probably forgot that `Observer` trait also needs to be changed. It would be like this:
> 
> ```rust
> pub trait Observer {
>     type Item;
>     type Err;
>     
>     fn next(&mut self, value: Self::Item);
>     fn error(&mut self, err: Self::Err);
>     fn complete(&mut self);
> }
> ```

`Observer` should change, but three is a problem blocked me. For now `Subject` emit item by mut ref, depends on Higher-Rank Trait Bounds,  below is the implementation:
```rust
impl<'a, Item, Err: Copy, O, U> RawSubscribable<Subscriber<O, U>>
  for LocalSubject<'a, SubjectMutRefValue<Item>, SubjectValue<Err>>
where
  O: for<'r> Observer<&'r mut Item, Err> + 'a,
  U: SubscriptionLike + Clone + 'static,
{
  local_subject_raw_subscribe_impl!(O, U);
}

```
is there a way i can bound with `Observer` like  this  ` O: for<'r> Observer<Item =&'r mut Item, Err> + 'a,`

**Comment by ishitatsuyuki** [comment]
On second thought, I think the `Observer` signature doesn't actually need to be changed.

It seems associated types is unnecessary and cause more issues like what you mentioned in this case. As the standard library [`Fn` trait types](https://doc.rust-lang.org/std/ops/trait.FnMut.html) also doesn't use associated types, `Observer` is best left as-is.

Though, you might want to merge `Observer{Next,Error,Complete}` into `Observer`, instead of using inheritance, as it would shorten the `impl` blocks.

**Comment by M-Adoo** [comment]
> ```rust
> trait Observable {
>    type Item;
>    type Err;
>    type Unsub; // RxJava style return values utilizes destructors
>    fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self::Err>>(self, observer: O) -> Self::Unsub;
> }
> ```
This is no correct, the method should not hold the generic `O`, when `observer` pass to upstream or downstream,  how can I specified the `O` type, in upstream?   For example:

```rust
struct ObservableFn<F>(F);

impl<F> Observable for ObservableFn<F>
where
  F: FnOnce(? = i want type `O` in method),
{
  fn actual_subscribe<O: Observer<Self::Item, Self::Err>>(
    self,
    observer: O,
  ) -> Self::Unsub {
    (self.0)(observer);
    ....
  }
}
```
So, `O` should be generic arg of trait or associated type, maybe this better:

```rust
trait Observable {
   type Item;
   type Err;
   type O: Observer<Self::Item, Self::Err>;
   type Unsub: SubscriptionLike;
   fn actual_subscribe(self, subscriber: O) -> Self::Unsub;
}
```

**Comment by ishitatsuyuki** [comment]
So well, we're left with two options: to revert to the current signature, or break ObservableFromFn with the upcoming change.

I still think the latter is preferable. The current signature, as you know, adds a lot of noise for implementors and users. And the only benefit it provides is the ability to use a closure for observable, which always can be substituted with a manual struct and implmentation of Observable.

Instead, you could make it work with associated types by using `Box<dyn Observer<...>>` for ObservableFromFn. This means that it will no longer be a zero-cost abstraction, but I think it's a good trade-off.

**Comment by M-Adoo** [comment]
Not only for ObservableFromFn, another case is  `Subject` emit item by mut ref, you can not give a correct lifetime.

```rust
impl<'a, Item, Err, O, U> Observable
  for LocalSubject<'a, SubjectMutRefValue<Item>, SubjectMutRefValue<Err>>
where
  O: for<'r> Observer<Item, &'r mut Err> + 'a,
  U: SubscriptionLike + Clone + 'static,
{
   type Item = &'? Item
   type Err = &'? Err
   type O = Box<...>

  fn actual_subscribe(self, observer: Self::O) -> Self::Unsub {
    ....
  }
}
```

Far more,  we should not bound Observer in Observable trait, but bound in its implementation.

And after a deeper think last night,  the subscribe method of Observable should take a observer from the user's perspective, but in  internal implementation `actual_subscribe` need to take a Subscriber that because sometimes we need something like Subscription. RxJava v1.x and rxjs accept Subscriber, and RxJava v2.x and v3.x  accept  something named observer, but that observer also implemented Disposable. 
```java
class SkipObserver<T> implements Observer<T>, Disposable
```

So, I will  refactor trait

``` rust 
pub trait RawSubscribable<Subscriber> {
  type Unsub: SubscriptionLike + 'static;
  fn raw_subscribe(self, subscriber: Subscriber) -> Self::Unsub;
}
```
to

```rust
pub trait Observable<O, U: SubscriptionLike> {
  type Unsub: SubscriptionLike + 'static;
  fn actual_subscribe(self, subscriber: Subscriber<O, U>) -> Self::Unsub;
}
```
And it's not a large work.

**Comment by helgoboss** [comment]
Sorry for the delay, I needed to experiment a little with a very minimal but naive Rx implementation to even understand what you guys are talking about ;) I would like to summarize things a bit. Can you tell me if I understood correctly?

As I understand it, from a user point of view, we were hoping to replace the following:

```rust
fn create_my_observable<O, U>() -> impl RawSubscribable<Subscriber<O, U>>
    where
        O: Observer<i32, ()>,
        U: SubscriptionLike + Clone + 'static
{
    observable::create(|mut subscriber| {
        subscriber.next(1);
        subscriber.next(2);
        subscriber.complete();
    })
}
```

With this:
```rust
fn create_my_observable() -> impl Observable<i32, ()> {
    observable::create(|mut observer| {
        observer.next(1);
        observer.next(2);
        observer.complete();
    })
}
```

Letting the **`Subscriber` vs `Observer`** and **type parameter vs. associated type** discussions aside for a moment: We wanted to achieve this by making `Observer` a type parameter of `actual_subscribe()` (as probably most other Rx implementations do it), not of `Observable`:

```rust
pub trait Observable<Item, Err> {
   type Unsub;

   fn actual_subscribe<O: Observer<Item, Err>>(self, observer: O) -> Self::Unsub;
}
```

But now we run into a problem: There are `Observable` implementations which need to know the concrete observer *at creation time*, not only at *subscribe time*. Example: `ObservableFromFn` created by `observable::create()`. This `struct` contains a closure `FnOnce(Observer)` which calls `next()`, `complete()` and `error()`, so `Observer` needs to be known when *constructing* the `struct`. With most other `Observable` implementations, this is not an issue because they need the observer in the  `actual_subscribe()` method only, which is called *after construction* (it is used at a later point in time by the consumer of this `Observable`).

Verdict: *If* rxRust should stay a zero-cost implementation even for `observable::create`, this is not possible. We don't have any other choice than making `Observer` a type parameter of `Observable` (very similar to status quo):

```rust
pub trait Observable<Item, Err, O: Observer<Item, Err>> {
    type Unsub;

    fn actual_subscribe(self, observer: O) -> Self::Unsub;
}
```

So all we could possibly achieve is this:

```rust
fn create_my_observable<O>() -> impl Observable<O>
    where
        O: Observer<i32, ()>
{
    observable::create(|mut observer| {
        observer.next(1);
        observer.next(2);
        observer.complete();
    })
}
```

... which is better than the first one, but still some noise. That noise it probably the minimum price that we have to pay for staying zero-cost.

If we would be okay with moving away from zero-cost, then we would keep the closure in `ObservableFromFn` as `FnOnce(Box<dyn Observer<Item, Err>>)` instead, problem solved (as @ishitatsuyuki suggested). AFAIK that would involve a heap allocation at creation time and a dynamic dispatch on each `next()`, `complete()` and `error()` call done by the source `Observable`.

Maybe rxRust is at a crossroad here: Should it stay zero-cost wherever possible or should it put more value on API ergonomics and parity with other Rx implementations?

@M-Adoo You mentioned this other issue with `&mut item` subject lifetimes if `Observer` is not a type parameter of `Observable`. This makes me wonder what is the actual use case for sending mutable stuff through Rx. I never felt the need for that, actually not even for passing immutable *references* ... but I'm probably not thinking broad enough here :) Still, wouldn't it be okay to resort to interior mutability in such cases?

**Comment by M-Adoo** [comment]
> Maybe rxRust is at a crossroad here: Should it stay zero-cost wherever possible or should it put more value on API ergonomics and parity with other Rx implementations?
 
In my opinion, we try to guarantee consistent usage, not implementation. 

> @M-Adoo You mentioned this other issue with &mut item subject lifetimes if Observer is not a type parameter of Observable. 

```rust
observable.subscribe(|event| {
     event.host.update_something(...);
]);
```

**Comment by M-Adoo** [comment]
After so much discussion, let us give a conclusion.  We hope to complete these refactorings list in #68  as soon as possible as first step.  But it didn't fully meet all our expectations，for `Observable` trait 
we hope it looks like this in the end: 
```rust
trait Observable {
   type Item;
   type Err;
   type Unsub;
   fn actual_subscribe<O: Observer<Item=Self::Item, Err=Self::Err>, U>(self,  subscriber: Subscriber<O, U>) -> Self::Unsub;
}
```
This will be consider after #68 finished, and think through all questions and confusions. @ishitatsuyuki @helgoboss

**Comment by M-Adoo** [comment]
Open a new topic about Observable trait. #71

**Comment by M-Adoo** [comment]
Open a new topic about Observable trait. #71

**Comment by M-Adoo** [comment]
closed by #68
